packages/rocketchat-lib/server/lib/processDirectEmail.js

```diff
@@ -0,0 +1,119 @@
+import {EmailReplyParser as reply} from 'emailreplyparser';
+
+RocketChat.processDirectEmail = function(email) {
+	function sendMessage(email) {
+		const message = {
+			ts: new Date(email.headers.date),
+			msg: email.body
+		};
+
+		if (!message.ts) {
+			message.ts = new Date();
+		}
+
+		if (message.msg && message.msg.length > RocketChat.settings.get('Message_MaxAllowedSize')) {
+			return false;
+		}
+
+		// reduce new lines in multiline message
+		message.msg = message.msg.split('\n\n').join('\n');
+
+		const user = RocketChat.models.Users.findOneByEmailAddress(email.headers.from, {
+			fields: {
+				username: 1,
+				name: 1
+			}
+		});
+		if (!user) {
+			// user not found
+			return false;
+		}
+
+		const prevMessage = RocketChat.models.Messages.findOneById(email.headers.mid, {
+			rid: 1,
+			u: 1
+		});
+		if (!prevMessage) {
+			// message doesn't exist anymore
+			return false;
+		}
+		message.rid = prevMessage.rid;
+
+		const room = Meteor.call('canAccessRoom', message.rid, user._id);
+		if (!room) {
+			return false;
+		}
+
+		// check mention
+		if (message.msg.indexOf(`@${ prevMessage.u.username }`) === -1) {
+			message.msg = `@${ prevMessage.u.username } ${ message.msg }`;
+		}
+
+		const roomInfo = RocketChat.models.Rooms.findOneById(message.rid, {
+			t: 1,
+			name: 1
+		});
+
+		// reply message link
+		let prevMessageLink = `[ ](${ Meteor.absoluteUrl().replace(/\/$/, '') }`;
+		if (roomInfo.t === 'c') {
+			prevMessageLink += `/channel/${ roomInfo.name }?msg=${ email.headers.mid }) `;
+		} else if (roomInfo.t === 'd') {
+			prevMessageLink += `/direct/${ prevMessage.u.username }?msg=${ email.headers.mid }) `;
+		} else if (roomInfo.t === 'p') {
+			prevMessageLink += `/group/${ roomInfo.name }?msg=${ email.headers.mid }) `;
+		}
+
+		message.msg = prevMessageLink + message.msg;
+
+		// info: message was sent by email
+		message.msg += '\n>Message sent by Email';
```

Comment:
It would be better to add a new property to the message object saying it came from email and render a letter icon by the username on UI